src/lib/utils/DY.Lib.SplitFunction.fst

```diff
@@ -0,0 +1,283 @@
+///
+/// The tag may be obtained in different ways,
+/// for example with the key usage (e.g. for a signature function, see DY.Core.Bytes.Type.usage),
+/// or because some messages are tagged
+/// (e.g. the state content in a state function or the signature input in signature function).
+/// How to obtain this tag is not always trivial,
+/// it is actually a core argument in the proof of
+/// why it is secure to deploy different sub-components in parallel.
+///
+/// The split function methodology can be used to implement horizontal composition:
+/// as long as we can find a way to disambiguate the composed protocols (using the tag),
+/// we can easily construct a global function implementing this horizontal composition.
+
+/// The parameters of the split function methodology.
+
+noeq type split_function_input_values = {
+  // Input type for the global function
+  tagged_data_t: Type;
+
+  // Functions to model set of tags
+  tag_set_t: Type;
+  tag_t: Type;
+  is_disjoint: tag_set_t -> tag_set_t -> prop;
+  tag_belong_to: tag_t -> tag_set_t -> bool;
+  cant_belong_to_disjoint_sets:
+    tag:tag_t -> tag_set_1:tag_set_t -> tag_set_2:tag_set_t -> Lemma
+    (~(tag_set_1 `is_disjoint` tag_set_2 /\ tag `tag_belong_to` tag_set_1 /\ tag `tag_belong_to` tag_set_2));
+
+
+  // Input type for the local functions
+  raw_data_t: Type;
+
+  // Output type for the local and global functions
+  output_t: Type;
+
+  // We can decode the global function input
+  // into an encoded tag and a local function input
+  decode_tagged_data: tagged_data_t -> Tot (option (tag_t & raw_data_t));
+
+  // Types for the local functions and the global function
+  local_fun: Type;
+  global_fun: Type;
+
+  default_global_fun: global_fun;
+
+  // Apply a local function to its input
+  apply_local_fun: local_fun -> raw_data_t -> output_t;
+  // Apply the global function to its input
+  apply_global_fun: global_fun -> tagged_data_t -> output_t;
+  // Create a global function
+  mk_global_fun: (tagged_data_t -> output_t) -> global_fun;
+  // Correctness theorem on creating and applying a global function
+  apply_mk_global_fun: bare:(tagged_data_t -> output_t) -> x:tagged_data_t -> Lemma
+    (apply_global_fun (mk_global_fun bare) x == bare x);
+}
+
+/// Do a global function contain some given local function with some set of tags?
+/// This will be a crucial precondition for the correctness theorem `local_eq_global_lemma`.
+
+val has_local_fun: func:split_function_input_values -> func.global_fun -> (func.tag_set_t & func.local_fun) -> prop
+let has_local_fun func gfun (tag_set, lfun) =
+  forall tagged_data.
+    match func.decode_tagged_data tagged_data with
+    | Some (tag, raw_data) ->
+      tag `func.tag_belong_to` tag_set ==> (func.apply_global_fun gfun tagged_data == func.apply_local_fun lfun raw_data)
+    | None -> True
+
+val find_local_fun: func:split_function_input_values -> list (func.tag_set_t & func.local_fun) -> func.tag_t -> option func.local_fun
```

Comment:
maybe add a comment on what this function does.
something along the lines of "checks if a given tag appears in a collection of tagged local functions and if so, returns the associated local function".
(is it important to mention that it returns the _first_ local function where the tag appears? in our use-cases it is probably unique since we have the disjointness property?)

Reply:
Done, I hope it is clear enough!

Reply:
Thanks!